### PR TITLE
Upgrade inline-c

### PR DESCRIPTION
**Pull request by junjihashimoto**
1, Fix quickcheck-error. See https://github.com/fpco/inline-c/pull/97 and https://github.com/fpco/inline-c/pull/99
2, Improve supporting c++template not to define each instance like `vector<tensor>` and `vector<int>`. See ATen/Type.hs.
3, Fix some CI-errors.

### CODE REVIEW

File: libtorch-ffi/src/ATen/Managed/Type/Tensor.hs [hunk 1]
```diff
@@ -2886,7 +2886,7 @@ tensor_multinomial_lbp
   :: ForeignPtr Tensor
   -> Int64
   -> CBool
-  -> ForeignPtr Generator
+  -> Ptr Generator
```
Comment (junjihashimoto) [inline libtorch-ffi/src/ATen/Managed/Type/Tensor.hs]: This code should be fixed.

File: .github/workflows/nix-macos.yml [hunk 2]
```diff
@@ -8,7 +8,7 @@ jobs:
     runs-on: macOS-latest
     steps:
     - uses: actions/checkout@v1
-    - uses: cachix/install-nix-action@v5
+    - uses: cachix/install-nix-action@v6
```
Comment (tscholak) [inline .github/workflows/nix-macos.yml]: 👍

File: nix/shared.nix [hunk 3]
```diff
@@ -91,10 +91,18 @@ let
...
+                            "inline-c-cpp"
+                            ../inline-c/inline-c-cpp
+                            { }
+                          )
+                          (old: {
+                              preConfigure = (old.preConfigure or "") + optionalString isDarwin ''
+                                sed -i -e 's/-optc-std=c++11//g' inline-c-cpp.cabal;
+                              '';
+                            }
+                          )
```
Comment (tscholak) [inline nix/shared.nix]: :+1:

**Comment by junjihashimoto** [comment]
This does not change  `Torch.Typed.Native`.
`StdTuple` is CppTuple's instance.

```
instance CppTuple2 (ForeignPtr (StdArray '(CBool,2))) where
  type A (ForeignPtr (StdArray '(CBool,2))) = CBool
  type B (ForeignPtr (StdArray '(CBool,2))) = CBool
  get0 v = cast1 (get0 :: Ptr (StdArray '(CBool,2)) -> IO CBool) v
  get1 v = cast1 (get1 :: Ptr (StdArray '(CBool,2)) -> IO CBool) v
```

CppTuple and Haskell-tuple becomes Castable's instance.

```
instance (CppTuple2 c, Castable a (A c), Castable b (B c)) => Castable (a,b) c where
  cast _ _ = undefined
  uncast t f = do
    t0 <- get0 t
    t1 <- get1 t
    t0' <- uncast t0 return
    t1' <- uncast t1 return
    f (t0',t1')
```

`Castable` converts `StdArray` to a Haskell tuple.

**Comment by tscholak** [comment]
thanks @junjihashimoto, I get it now

**Comment by junjihashimoto** [comment]
Thank you for your reviewing.

**Pull request merged**